### PR TITLE
fix ipfs upload

### PR DESCRIPTION
**Pull request by yann300**
fix https://github.com/ethereum/remix-ide/issues/2540

### CODE REVIEW

File: src/lib/publishOnIpfs.js [hunk 1]
```diff
@@ -27,11 +27,21 @@ module.exports = (contract, fileManager, cb, ipfsVerifiedPublishCallBack) => {
 
   async.eachSeries(Object.keys(metadata.sources), function (fileName, cb) {
     // find hash
-    var hash
+    var hash = null
```
Comment (Aniket-Engg) [inline src/lib/publishOnIpfs.js]: Can we use `let` here?

File: src/lib/publishOnIpfs.js [hunk 2]
```diff
@@ -27,11 +27,21 @@ module.exports = (contract, fileManager, cb, ipfsVerifiedPublishCallBack) => {
...
-      hash = metadata.sources[fileName].urls[1].match('dweb:/ipfs/(.+)')[1]
+      // we try extract the hash defined in the metadata.json
+      // in order to check if the hash that we get after publishing is the same as the one located in metadata.json
+      // if it's not the same, we throw "hash mismatch between solidity bytecode and uploaded content"
+      // if we don't find the hash in the metadata.json, the check is not done.
+      //
+      // TODO: refactor this with publishOnSwarm
+      if (metadata.sources[fileName].urls) {
+        metadata.sources[fileName].urls.forEach(url => {
+          if (url.indexOf('ipfs') !== -1) hash = url.match('dweb:/ipfs/(.+)')[1]
```
Comment (Aniket-Engg) [inline src/lib/publishOnIpfs.js]: Can you consider using `.includes` instead of `.indexOf` on all the places

File: src/lib/publishOnIpfs.js [hunk 3]
```diff
@@ -94,7 +104,7 @@ module.exports = (contract, fileManager, cb, ipfsVerifiedPublishCallBack) => {
 async function ipfsVerifiedPublish (content, expectedHash, cb) {
   try {
     const results = await severalGatewaysPush(content)
-    if (results !== expectedHash) {
+    if (expectedHash !== null && results !== expectedHash) {
```
Comment (Aniket-Engg) [inline src/lib/publishOnIpfs.js]: Can we do `(expectedHash&& results !== expectedHash)` for all similar places?

File: src/lib/publishOnSwarm.js [hunk 4]
```diff
@@ -20,11 +20,21 @@ module.exports = (contract, fileManager, cb, swarmVerifiedPublishCallBack) => {
 
   async.eachSeries(Object.keys(metadata.sources), function (fileName, cb) {
     // find hash
-    var hash
+    var hash = null
```
Comment (Aniket-Engg) [inline src/lib/publishOnSwarm.js]: Can we use `let` here?

**Comment by yann300** [comment]
ok added the changes @Aniket-Engg

**Pull request merged**